aiohttp_cache/middleware.py

```diff
@@ -1,25 +1,52 @@
-from typing import Any, TypeVar
+import functools
+
+from typing import Awaitable, Callable, Type, Union
 
 from aiohttp import web
+from aiohttp.abc import AbstractView, StreamResponse
+from aiohttp.web_request import Request
+from aiohttp.web_response import Response
+
 
+_WebHandler = Callable[[Request], Awaitable[StreamResponse]]
+HandlerType = Union[_WebHandler, Type[AbstractView]]
 
-HandlerType = TypeVar("HandlerType", bound=Any)
-T = TypeVar("T", bound=Any)
 
+def get_original_handler(
+    handler: Union[HandlerType, functools.partial[HandlerType]]
+) -> HandlerType:
+    """Return the original handler object.
 
-def get_original_handler(handler: HandlerType) -> HandlerType:
+    In case the handler was provided as functools.partial or
+    handler is hidden under chain of middlewares
+    we need to extract the original handler object, in
+    order to inspect if it is intended to be cached or not.
+    """
     if hasattr(handler, "cache_enable"):
-        return handler
+        return handler  # type: ignore
     elif hasattr(handler, "keywords"):
-        return get_original_handler(handler.keywords["handler"])
+        try:
+            return get_original_handler(
+                handler.keywords["handler"]  # type: ignore
+            )
+        except KeyError:
+            # handle the case when the handler type is functools.partial
+            return handler.func  # type: ignore
```

Comment:
In case this is a partial handler (not a middleware),  then we return exactly the handler function (so it will have caching metadata)